### ISSUE TITLE
Скачивание-загрузка должны выполняться по хэшу

### ISSUE THREAD

**Issue by pansa** [tmp1test/t2#3]
По заданию запросы на скачивание и удаление происходят по хэшу файла.
В вашей реализации клиенту необходимо помимо хэша знать остальную часть имени?
Такой подход трудноприменим в микросервисной среде, ведь скачивать/удалять файл могут совершенно другие клиенты, нежели те, что его загрузили. 

Также, как поведет себя ваш сервис, если один и тот же файл загружается много-много раз?

**Comment by tmp1test** [comment]
@pansa 
Мой посыл был в том, что два клиента могут загрузить один и тот же файл. При этом, если один удалит свой файл, файл второго клиента останется в хранилище и он может продолжать с ним работу.
В ответ на загрузку файла выдается ответ с хешем, который используется при скачаивании/удалении файла. Для скачивания/удаления файла можно передать данный хеш другому пользователю.

Могу переделать вечером на вариант без возможности дублирования файлов, если требуется)

**Comment by pansa** [comment]
> Мой посыл был в том, что два клиента могут загрузить один и тот же файл. При этом, если один удалит свой файл, файл второго клиента останется в хранилище и он может продолжать с ним работу.

Ваш вариант понятен, но по условию задачи это не тробовалось. Формат же хранения, описанный в задании, исключал хранение дублей. 

> Для скачивания/удаления файла можно передать данный хеш другому пользователю.
Именно в этом и проблема =)

>Могу переделать вечером на вариант без возможности дублирования файлов, если требуется)
Если у вас есть время и желание, то будет интересно. Но вам придется много чего поменять.

**Comment by tmp1test** [comment]
@pansa 
На самом деле этот вопрос исходя из описании не до конца ясен )
Я бы предложил внести в описание правку, в которой явно будет указано, что дублей быть не должно. Так будет яснее следующим кандидатам

Исправил, теперь дубли невозможны. В качестве хеша используется sha256

**Comment by pansa** [comment]
Вы "откусили" суффиксы у загружаемых файлов. 
Но что будет, если два клиента одновременно начнут загружать одинаковый файл?  
Если у одного из них оборвется соединение? 
Также, в процессе загрузки файла вы сразу разместили его в сторадже, еще ДО его окончательной загрузки, а это означает, что если другой клинет в это время начнет скачивать файл, он получит "огрызок" или ошибку.

**Comment by tmp1test** [comment]
@pansa 
понял вас, предлагаю ввести предхранилище для файлов
перед записью в него клиент ждет какой-то таймаут, возможно, что файл уже будет загружен другим клиентом.
отписал об этом также в другом ишью

устроит ли такой вариант?

вечером исправлю

**Comment by pansa** [comment]
Нет. Если честно, то этот вариант очень не нравится, т.к превносит еще больше проблем  =)  
Тем более, что такая "защита" не сработает, ведь нет никаких гарантий, что спустя таймаут ситуация с двумя клиентами не повториться вновь. 

Решение очень простое, подумайте. По сути нам надо защититься от того, чтобы два потока работали с одним и тем же файлом в один и тот же момент.

**Comment by tmp1test** [comment]
@pansa 
добавил предхранилище файлов и мютекс для него
теперь сразу несколько клиентов смогут загрузить один и тот же файл
